sematic/graph.py

```diff
@@ -0,0 +1,560 @@
+        runs of all ancestors are set to CREATED.
+
+        If reset_from is None, only FAILED, NESTED_FAILED, and CANCELED runs
+        will be reset.
+
+        Parameters
+        ----------
+        storage: Storage
+            The storage class to retrieve artifact values and set future.value
+            and future.kwargs appropriately.
+
+        reset_from: Optional[str]
+            Force reset other runs than only failed ones.
+
+        Returns
+        -------
+        Tuple[List[AbstractFuture], Dict[str, Dict[str, Artifact]], Dict[str,
+        Artifact]]
+            A tuple whose first element is a list of cloned futures, grouped by
+            nested layers (outermost first), and sorted by reverse execution
+            order within each layer. The second element is a mapping of future
+            IDs to input artifacts. The third element is a mapping of future IDs
+            to output artifacts.
+
+        """
+        value_by_artifact_id: Dict[str, Any] = {}
+
+        futures_by_original_id: Dict[RunID, AbstractFuture] = {}
+        input_artifacts: Dict[RunID, Dict[str, Artifact]] = defaultdict(dict)
+        output_artifacts: Dict[RunID, Artifact] = {}
+
+        skip_run_ids, reset_run_ids = (
+            ([], []) if reset_from is None else self._get_skip_reset_run_ids(reset_from)
+        )
+
+        # run order guarantees parents and upstream come first
+        # This is necessary because we want upstream cloned futures
+        # to be created before downstreams so that the appropriate
+        # kwargs can be built
+        run_ids_by_execution_order = self._sorted_run_ids_by_layer(
+            run_sorter=self._execution_order
+        )
+
+        def _get_artifact_value(artifact: Artifact) -> Any:
+            if artifact.id not in value_by_artifact_id:
+                value_by_artifact_id[artifact.id] = get_artifact_value(
+                    artifact, storage
+                )
+
+            return value_by_artifact_id[artifact.id]
+
+        def _get_run_inputs(
+            run_id: RunID,
+        ) -> Tuple[Dict[str, Any], Dict[str, Artifact]]:
+            kwargs: Dict[str, Any] = {}
+
+            input_edges = self._edges_by_destination_id[run_id]
+
+            run_input_artifacts: Dict[str, Artifact] = {}
+
+            for input_edge in input_edges:
+                if input_edge.destination_name is None:
+                    raise RuntimeError("Input edge misses destination name")
+
+                value = None
+
+                if input_edge.artifact_id is not None:
+                    artifact = self._artifacts_by_id[input_edge.artifact_id]
+                    run_input_artifacts[input_edge.destination_name] = artifact
+                    value = _get_artifact_value(artifact)
+
+                if input_edge.source_run_id is not None:
+                    kwargs[input_edge.destination_name] = futures_by_original_id[
```

Comment:
maybe add a comment about why we prefer to set the arg with the future instead of with the value directly? I assume it's so that we can maintain the in-memory links between the futures but it would be good to call that out.

Reply:
Done.